File: 2026-protocol/nextcloud-file-browser/storage.go
package main

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
)

// Credential stores Nextcloud access credentials for an identity
type Credential struct {
	NextcloudURL string `json:"nextcloud_url"`
	Username     string `json:"username"`
	AppPassword  string `json:"app_password"`
}

// CredentialStore manages stored credentials
type CredentialStore struct {
	Credentials map[string]*Credential `json:"credentials"`
	configDir   string
}

func getConfigDir() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(home, ".config", "gradesta", "nextcloud-file-browser"), nil
}

// loadCredentials loads the credential store from disk
func loadCredentials() (*CredentialStore, error) {
	configDir, err := getConfigDir()
	if err != nil {
		return nil, err
	}

	store := &CredentialStore{
		Credentials: make(map[string]*Credential),
		configDir:   configDir,
	}

	credPath := filepath.Join(configDir, "credentials.enc")
	if _, err := os.Stat(credPath); os.IsNotExist(err) {
		return store, nil
	}

	// Read encrypted file
	encData, err := os.ReadFile(credPath)
	if err != nil {
		return nil, fmt.Errorf("failed to read credentials: %w", err)
	}

	// Load encryption key
	key, err := loadOrCreateKey(configDir)
	if err != nil {
		return nil, fmt.Errorf("failed to load key: %w", err)
	}

	// Decrypt
	data, err := decrypt(encData, key)
	if err != nil {
		return nil, fmt.Errorf("failed to decrypt credentials: %w", err)
	}

	// Parse JSON
	if err := json.Unmarshal(data, store); err != nil {
		return nil, fmt.Errorf("failed to parse credentials: %w", err)
	}

	return store, nil
}

// Save saves the credential store to disk
func (s *CredentialStore) Save() error {
	if err := os.MkdirAll(s.configDir, 0700); err != nil {
		return fmt.Errorf("failed to create config dir: %w", err)
	}

	// Load encryption key
	key, err := loadOrCreateKey(s.configDir)
	if err != nil {
		return fmt.Errorf("failed to load key: %w", err)
	}

	// Serialize to JSON
	data, err := json.MarshalIndent(s, "", "  ")
	if err != nil {
		return fmt.Errorf("failed to serialize credentials: %w", err)
	}

	// Encrypt
	encData, err := encrypt(data, key)
	if err != nil {
		return fmt.Errorf("failed to encrypt credentials: %w", err)
	}

	// Write file
	credPath := filepath.Join(s.configDir, "credentials.enc")
	if err := os.WriteFile(credPath, encData, 0600); err != nil {
		return fmt.Errorf("failed to write credentials: %w", err)
	}

	return nil
}

// Get retrieves credentials for an identity
func (s *CredentialStore) Get(identity string) (*Credential, bool) {
	cred, ok := s.Credentials[identity]
	return cred, ok
}

// Set stores credentials for an identity
func (s *CredentialStore) Set(identity string, cred *Credential) {
	s.Credentials[identity] = cred
}

// loadOrCreateKey loads or generates the encryption key
func loadOrCreateKey(configDir string) ([]byte, error) {
	keyPath := filepath.Join(configDir, "secret.key")

	if data, err := os.ReadFile(keyPath); err == nil && len(data) == 32 {
		return data, nil
	}

	// Generate new key
	key := make([]byte, 32)
	if _, err := rand.Read(key); err != nil {
		return nil, err
	}

	if err := os.MkdirAll(configDir, 0700); err != nil {
		return nil, err
	}

	if err := os.WriteFile(keyPath, key, 0600); err != nil {
		return nil, err
	}

	return key, nil
}

// encrypt encrypts data using AES-256-GCM
func encrypt(plaintext, key []byte) ([]byte, error) {
	// Derive key using SHA-256 to ensure correct length
	hash := sha256.Sum256(key)
	derivedKey := hash[:]

	block, err := aes.NewCipher(derivedKey)
	if err != nil {
		return nil, err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}

	nonce := make([]byte, gcm.NonceSize())
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
		return nil, err
	}

	return gcm.Seal(nonce, nonce, plaintext, nil), nil
}

// decrypt decrypts data using AES-256-GCM
func decrypt(ciphertext, key []byte) ([]byte, error) {
	// Derive key using SHA-256 to ensure correct length
	hash := sha256.Sum256(key)
	derivedKey := hash[:]

	block, err := aes.NewCipher(derivedKey)
	if err != nil {
		return nil, err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}

	nonceSize := gcm.NonceSize()
	if len(ciphertext) < nonceSize {
		return nil, fmt.Errorf("ciphertext too short")
	}

	nonce, ciphertext := ciphertext[:nonceSize], ciphertext[nonceSize:]
	return gcm.Open(nil, nonce, ciphertext, nil)
}

File: 2026-protocol/nextcloud-connector/src/storage.rs
//! Credential storage with AES-256-GCM encryption

use aes_gcm::{
    aead::{Aead, KeyInit},
    Aes256Gcm, Nonce,
};
use anyhow::{anyhow, Context, Result};
use rand::RngCore;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;

/// Stored credentials for a Nextcloud account
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Credential {
    pub nextcloud_url: String,
    pub username: String,
    pub app_password: String,
}

/// Credential store
#[derive(Debug, Default, Serialize, Deserialize)]
pub struct CredentialStore {
    pub credentials: HashMap<String, Credential>,
    #[serde(skip)]
    config_dir: PathBuf,
}

impl CredentialStore {
    /// Get config directory path
    /// Uses /data if available (Docker volume mount), otherwise falls back to user config dir
    fn get_config_dir() -> Result<PathBuf> {
        // Check for Docker volume mount first
        let data_dir = PathBuf::from("/data");
        if data_dir.exists() {
            return Ok(data_dir);
        }

        // Fall back to user config directory
        let home = dirs::config_dir().ok_or_else(|| anyhow!("No config directory"))?;
        Ok(home.join("gradesta").join("nextcloud-connector"))
    }

    /// Load credentials from disk (decrypted)
    pub fn load() -> Result<Self> {
        let config_dir = Self::get_config_dir()?;
        let cred_path = config_dir.join("credentials.enc");

        if !cred_path.exists() {
            return Ok(Self {
                credentials: HashMap::new(),
                config_dir,
            });
        }

        // Read encrypted file
        let enc_data = fs::read(&cred_path).context("Failed to read credentials")?;

        // Load encryption key
        let key = load_or_create_key(&config_dir)?;

        // Decrypt
        let data = decrypt(&enc_data, &key)?;

        // Parse JSON
        let mut store: Self = serde_json::from_slice(&data).context("Failed to parse credentials")?;
        store.config_dir = config_dir;

        Ok(store)
    }

    /// Save credentials to disk (encrypted)
    pub fn save(&self) -> Result<()> {
        fs::create_dir_all(&self.config_dir).context("Failed to create config dir")?;

        // Load encryption key
        let key = load_or_create_key(&self.config_dir)?;

        // Serialize to JSON
        let data = serde_json::to_vec_pretty(self)?;

        // Encrypt
        let enc_data = encrypt(&data, &key)?;

        // Write file
        let cred_path = self.config_dir.join("credentials.enc");
        fs::write(&cred_path, enc_data).context("Failed to write credentials")?;

        // Set permissions (Unix only)
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            fs::set_permissions(&cred_path, fs::Permissions::from_mode(0o600))?;
        }

        Ok(())
    }

    /// Get credentials for an identity
    pub fn get(&self, identity: &str) -> Option<&Credential> {
        self.credentials.get(identity)
    }

    /// Set credentials for an identity
    pub fn set(&mut self, identity: &str, cred: Credential) {
        self.credentials.insert(identity.to_string(), cred);
    }
}

/// Load or create encryption key
fn load_or_create_key(config_dir: &PathBuf) -> Result<[u8; 32]> {
    let key_path = config_dir.join("secret.key");

    if let Ok(data) = fs::read(&key_path) {
        if data.len() == 32 {
            let mut key = [0u8; 32];
            key.copy_from_slice(&data);
            return Ok(key);
        }
    }

    // Generate new key
    let mut key = [0u8; 32];
    rand::thread_rng().fill_bytes(&mut key);

    fs::create_dir_all(config_dir)?;
    fs::write(&key_path, &key)?;

    // Set permissions (Unix only)
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        fs::set_permissions(&key_path, fs::Permissions::from_mode(0o600))?;
    }

    Ok(key)
}

/// Encrypt data using AES-256-GCM
fn encrypt(plaintext: &[u8], key: &[u8; 32]) -> Result<Vec<u8>> {
    // Derive key using SHA-256
    let hash = Sha256::digest(key);
    let cipher = Aes256Gcm::new_from_slice(&hash).context("Failed to create cipher")?;

    // Generate random nonce
    let mut nonce_bytes = [0u8; 12];
    rand::thread_rng().fill_bytes(&mut nonce_bytes);
    let nonce = Nonce::from_slice(&nonce_bytes);

    // Encrypt
    let ciphertext = cipher
        .encrypt(nonce, plaintext)
        .map_err(|_| anyhow!("Encryption failed"))?;

    // Prepend nonce to ciphertext
    let mut result = Vec::with_capacity(12 + ciphertext.len());
    result.extend_from_slice(&nonce_bytes);
    result.extend_from_slice(&ciphertext);

    Ok(result)
}

/// Decrypt data using AES-256-GCM
fn decrypt(ciphertext: &[u8], key: &[u8; 32]) -> Result<Vec<u8>> {
    if ciphertext.len() < 12 {
        return Err(anyhow!("Ciphertext too short"));
    }

    // Derive key using SHA-256
    let hash = Sha256::digest(key);
    let cipher = Aes256Gcm::new_from_slice(&hash).context("Failed to create cipher")?;

    // Extract nonce and ciphertext
    let nonce = Nonce::from_slice(&ciphertext[..12]);
    let encrypted = &ciphertext[12..];

    // Decrypt
    cipher
        .decrypt(nonce, encrypted)
        .map_err(|_| anyhow!("Decryption failed"))
}
